### ISSUE TITLE
Implement finetune on Families In The Wild

### ISSUE THREAD

**Issue by vitalwarley** [vitalwarley/research#18]
- Finetune pretrained model from [vuvko/fitw2020](https://github.com/vuvko/fitw2020)
    - [ ] #20
    - [x] #21
    - [x] #22

**Comment by vitalwarley** [comment]
Shadrikov (2020) used 

- `retinaface_r50_v1` from insightface to re-detect and align faces. They're [here](https://github.com/vuvko/fitw2020/tree/master/val-faces-det).
- `arcface_r100_v1`, also from insightface, which I am [trying to train](https://github.com/vitalwarley/research/issues/9#issuecomment-1106763746), was fed these re-detected and aligned images. The _baseline_ below is the result of this approach.

![2022-04-22-164921_839x711_scrot](https://user-images.githubusercontent.com/6365065/164784143-8e115824-a8d5-43f7-a817-257797856fb8.png)

- The next step was to fine-tune the whole model on the train set **with a simple classification layer**.
- The final experiment was to add an L2 normalization of the embeddings and retrain the original pre-trained model.
  - From what I understood, this normalization occurred before the classification layer. That is, we have here a pre-trained model + classification layer + normalization layer.
- The final threshold selected for the cosine distance between images was that which gave a False Positive Rate of 0.2

**Comment by vitalwarley** [comment]
My first try on getting the baseline.

```
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
     Validate metric           DataLoader 0
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
         val_acc                    0.5
         val_auc             0.54376220703125
──────────────────────────────────────────
```

AUC should have been 0.7183, so there is something wrong with my approach.

**Comment by vitalwarley** [comment]
I thought that the weights weren't being loaded. I used `--ckpt-path` instead `--weights`, but nothing changed. Below I list and describe where I think the error is.

### KinshipDataModule or FamiliesDataset;

It may be that the pairs are wrong. In the paper, Shadrikov says

> We sampled 5 000 positive and 5 000 negative pairs selected uniformly between all families from the validation set using algorithm 1.

Which seems to be done here

https://github.com/vitalwarley/research/blob/eb90479aa8b4c65dbe60f3bc1f6d7a26438db5d5/dataset.py#L237-L261

and here

https://github.com/vitalwarley/research/blob/eb90479aa8b4c65dbe60f3bc1f6d7a26438db5d5/dataset.py#L269-L286

However, in the [original implementation](https://github.com/vuvko/fitw2020/blob/27cd4ec8a54257a43629bde0504d784391ed67a0/verification.py#L151-L172), the code that uses the above scheme is commented. Instead, they simply used `fitw/val_pairs.csv`, which also has 10k lines.

I will try to use this .csv directly. If it is not the cause, there are other alternatives that I will check: image transformations,  `validation_step`, where I compute pair similarity, and `validation_epoch_end`, where I compute accuracy and AUC from similarities.

**Comment by vitalwarley** [comment]
I tried with 43018e24cfb71b9a3018e834b312a8e519d9c7ba, but it didn't work.

Next steps

- Check `validation_step` 
- Check `validation_epoch_end` 

I don't think the problem is in some missing or not transformation. The only I use for validation is `ToTensor`, which converts the array to a tensor and to CHW storage order. In `fitw2020/verification.py`, there is no pre-processing

```python
                img = mx.img.imread(str(im_path)).transpose((2, 0, 1)).expand_dims(0).astype(np.float32)
                batch = mx.io.DataBatch([img])
                self.model.forward(batch, is_train=False)
                self.embeddings[im_path] = self.model.get_outputs()[0][0].asnumpy()

```

**Comment by vitalwarley** [comment]
The cosine similarity I use is [this](https://pytorch.org/docs/stable/generated/torch.nn.functional.cosine_similarity.html), which is not different from Shadrikov original implementation

```python
def norm(emb):
    return np.sqrt(np.sum(emb ** 2))


def cosine(emb1, emb2):
    return np.dot(emb1, emb2) / (norm(emb1) * norm(emb2))

```

Just in case, I tested both

```python

[ins] In [18]: cosine(inp1, inp2.transpose())
Out[18]: array([[-0.03555923]])

[ins] In [19]: cosine_similarity(torch.from_numpy(inp1), torch.from_numpy(inp2))
Out[19]: tensor([-0.0356], dtype=torch.float64)

```

**Comment by vitalwarley** [comment]
Every pair is having a high similarity -- minimum of 0.78, maximum of 0.998 -- with my code and model. With the original implementation, however, I get -0.286 and 0.82 for min and max similarity, respectively. This is evidence that the problem is not on `validation_epoch_end`, but before it -- somewhere along the way from image reading to the `validation_step` itself.

**Comment by vitalwarley** [comment]
I run my model and the pretrained resnet101 that Shadrikov uses on a pair and obtained these similarities

```
Image 1: similarity between embeddings from my model vs those from insightface: -0.014783437363803387
Image 2: similarity between embeddings from my model vs those from insightface: 0.016696179285645485
```

The expected similarity is 1.0, because both models were trained on the same dataset with basically the same config. However, it seems that my model computing an orthogonal embedding...

How I did it

 ```python
    sample = pair_list[0]

    # my model
    img1, img2 = read_img_as_tensor(sample[0]), read_img_as_tensor(sample[1])
    embeddings1, _ = model(img1)
    embeddings2, _ = model(img2)

    # shadrikov model
    ctx = mx.cpu(0)
    mx_model = CompareModel(ctx=ctx)
    mx_embeddings1 = mx_model.get_embedding(sample[0])
    mx_embeddings2 = mx_model.get_embedding(sample[1])

    mx_embeddings1 = torch.from_numpy(mx_embeddings1).float()
    mx_embeddings2 = torch.from_numpy(mx_embeddings2).float()

    print("Image 1: similarity between embeddings from my model vs those from insightface: ", F.cosine_similarity(embeddings1, mx_embeddings1).item())
    print("Image 2: similarity between embeddings from my model vs those from insightface: ", F.cosine_similarity(embeddings2, mx_embeddings2).item())

```

and I can see nothing wrong.

--- 

After, I did that to all pairs and got 

```python
Image 1: similarity between embeddings from my model vs those from insightface: 0.0052 (+/- 0.0444)
Image 2: similarity between embeddings from my model vs those from insightface: 0.0043 (+/- 0.0442)
```

**Comment by vitalwarley** [comment]
Well, I don't know if the results above prove something... With backbone.pth from insightface, the results are similar to the above, that is, they are somewhat orthogonal. However, with backbone.pth and my weights, the results are also orthogonal between themselves. I think I am missing something.

**Comment by vitalwarley** [comment]
For Shadrikov, if I scale image to [0, 1], the results change -- all similarities become ~1. With my model, the input is originally in [0, 1], and changing to [0, 255] doesn't make a difference. It is expected, I think, because I trained in that range.

**Comment by vitalwarley** [comment]
Failed attemps

- I exported the model to .onnx and put it on Shadrikov script. Same bad results.
- I normalized the images just like I did for training (`(img - 0.5) / 0.5)`). Same bad results.
  - On my code and on Shadrikov script with the .onnx model.

**Comment by vitalwarley** [comment]
Failled attempts

- My model .onnx on LFW gives 0.6375 AUC.
- Insightface arcface_r100_v1 (mxnet; baseline for Shadrikov) gives 0.9994 AUC.

**Comment by vitalwarley** [comment]
I was going to do a summary of what I tried, but I think I found the cause of the problem.

My aim is to reproduce Shadrikov baseline (blue line [here](https://github.com/vitalwarley/research/issues/18#issuecomment-1106821035)). He used resnet101 (using mxnet; model named `arcface_r100_v1`), which I wrongly thought that it was trained on **MS1M_v3** (face recognition dataset; [link)](https://github.com/deepinsight/insightface/tree/master/recognition/_datasets_#ms1m-retinaface). Actually, the model was trained on **MS1M_v2: MS1M-ArcFace** (R100 (mxnet) [here](https://github.com/deepinsight/insightface/tree/master/model_zoo#list-of-models-by-mxnet-and-paddlepaddle)).  I don't know the difference between the datasets apart from number of classes and images. It should be something important to not help with the baseline.

I was mistaken because I used [insightface/recognition/arcface_torch](https://github.com/deepinsight/insightface/tree/master/recognition/arcface_torch) to guide my implementation. [There they use MS1M_v3](https://github.com/deepinsight/insightface/tree/master/recognition/arcface_torch#download-datasets-or-prepare-datasets). The model used by Shadrikov is a mxnet model, therefore I should also have used [insightface/recognition/arcface_mxnet](https://github.com/deepinsight/insightface/tree/master/recognition/arcface_mxnet#model-training) to guide me.

**Comment by vitalwarley** [comment]
It seems that the [dataset is not the problem](https://github.com/vitalwarley/research/issues/9#issuecomment-1120193092).

## Summary

- My aim
  - Get a 0.7183 AUC on Families In The Wild (kinship) dataset with **ResNet101**  as a feature extractor, which was previously trained on the MS1M_v2 (general face) dataset. _This is Shadrikov's baseline._
    - Shadrikov used the pretrained mxnet model from insightface.
- What I did: 
  - I trained a resnet101 with ArcFace loss on MS1M_v2.
    -  I reimplemented it with [timm](https://github.com/rwightman/pytorch-image-models) in PyTorch + PyTorch Ligthning.
- My problem
  - My accuracy and norms on LFW, AGEDB, CFP are similar to those from insightface.
  - My  Area Under ROC (AUC) **is far from similar** to that of Shadrikov.

## Discussion

At first, if we consider the validation metrics (accuracy), it makes no sense to be a problem with the model I trained. I checked this position and obtained some evidence against this hypothesis. I converted insightface original pytorch model and my trained model to ONNX. After, I used both in the Shadrikov's own script with ONNX Runtime. The results are below.

- I get AUC=0.99 on LFW with `arcface_r100_v1.mxnet` (Shadrikov original MXNET model)
- `arcface_r100_v1.onnx` (insightface original pytorch model converted to ONNX) gives only AUC=0.63.
-  My model as .onnx gives me only AUC=0.65. 

Model problem? I don't think so. Image processing problem? Maybe, because another thing I did was to get accuracy on LFW using `val_shadrikov.py` (using .ONNX) and `val.py` (using my original implementation).

- `val_shadrikov.py` gives accuracy = 0.60
- `val.py` gives accuracy = 0.99, as it should be. 
 
The difference between the two validation approaches are basically image preprocessing (maybe) and model framework (onnx and pytorch, respectively).  

## TODO

- Investigate image processing (dataset and datamodule).
- Check ONNX model inside `model.py` validation step.
- Investigate embeddings

This last item I have slowly started, and below is a batch embedding from LFW. I will check how to add proper labels.

![2022-05-08-221939_1878x982_scrot](https://user-images.githubusercontent.com/6365065/167325738-952f493f-6c9e-498d-b5c0-bea6da710fdb.png)

**Comment by vitalwarley** [comment]
- [ ] Investigate the difference between lfw implementation on datadaset.py and on val_shadrikov.py

**Comment by vitalwarley** [comment]
Both accuracies are ~0.5 because the similarities are low (< 0.5, which is the threshold I used). See them below for both models

![CompareModelSimilarity](https://user-images.githubusercontent.com/6365065/172029178-27c858a9-0751-4045-bfca-7dc11ff2a30e.png)

Shadrikov didn't report accuracy anywhere but talked about the binarization threshold scheme (section III.D), so I think we shouldn't aim for accuracy here, albeit its importance.

**Comment by vitalwarley** [comment]
On `KinshipDataModule`, which I use in the "datamodule" scheme, there are the following image processing ops

- FamiliesDataset with passed val_transform below

```

**Comment by vitalwarley** [comment]
> Shadrikov didn't report accuracy anywhere but talked about the binarization threshold scheme (section III.D), so I think we shouldn't aim for accuracy here, albeit its importance.

Actually, it was reported...

> We should note that though our approach scores first on average it is not by a far margin and mostly due to our great performance on sibling pairs. Having an 0.78 average accuracy at best, automatic kinship recognition still needs to be improved to be considered for usage in real-world applications. For the reference, the best face verification models perform with FPR around 10−6 [7].
> Section 4.C

Before it we also have

> We needed to binarize our predictions to submit for test verification on track 1. We chose threshold such that we had true positive rate (TPR) of 0.75 on our sampled validation images for our first submission (pretrained in table I). Next submissions were tested with different thresholds and between several strategies, the one that showed the best average performance was to choose the threshold such that the method would have a false positive rate (FPR) of 0.2. Other entries in table I are provided with that strategy of choosing the binarization threshold.
> Section 4.C

![2022-06-05-082708_1323x382_scrot](https://user-images.githubusercontent.com/6365065/172048247-d504e12c-fef3-4f6e-8eef-b7ceb86a3b3d.png)

This table shows the track 1 accuracy results between kinship types as well as the average accuracy for all strategies. The threshold seems to be 0.05, as we can see below, but the accuracy is lower than the expected (0.67 vs 0.70 in the table above). Why?

![mxnet_threshold_vs_accuracy](https://user-images.githubusercontent.com/6365065/172052006-1d1278df-6166-466d-bd38-d1488820e0d0.png)

I computed on all pairs. Soon I will do it for each kinship relation.

**Comment by vitalwarley** [comment]
> ## TODO
> ...
> * [ ]  Investigate difference between datamodule and no datamodule approaches for FIW

I wasn't using the right `mining_strategy` for the datamodule approach. It should have been "baseline" to be the same scheme as the raw approach (no datamodule). I changed it, but the results didn't change (0.54 and 0.52, for torch and mxnet, respectively).

Also, I noticed that embeddings from datamodule vs raw are too different. More evidence of some error in image processing ops

## Raw

![CompareModelEmbeddings](https://user-images.githubusercontent.com/6365065/172066345-b31e8845-3734-4925-b4a7-5f0509372805.png)

## Datamodule

![CompareModelEmbeddings](https://user-images.githubusercontent.com/6365065/172066350-e4b4208d-012e-4ff1-a4e7-91157f1ad452.png)

**Comment by vitalwarley** [comment]
I found a bug in `KinshipDataModule` validation transforms. The `ToTensor` transform didn't scale the image because it has `dtype = np.float64`. After fixing it, we now have the same results (accuracy and AUC) as in the RAW approach.

![CompareModelEmbeddings](https://user-images.githubusercontent.com/6365065/172067283-e63c9489-cf2f-4be3-a36c-fe08987db1b5.png)

> ## TODO (https://github.com/vitalwarley/research/issues/18#issuecomment-1123844210)
> [ ] Investigate difference between datamodule and no datamodule approaches for FIW

**Comment by vitalwarley** [comment]
> TODO (https://github.com/vitalwarley/research/issues/18#issuecomment-1123844210)
> - [ ] Investigate difference between PairDataset (fiw) and EvalPretrainDataset (lfw)

I think the only difference was that one mentioned in the previous comment.

**Comment by vitalwarley** [comment]
Updated results below. AUC results on FIW should be 0.7183.

## Datamodule 

- ArcFace R100 ONNX (my model): 0.6208
- ArcFace R100 MXNET: 0.7183

## Raw

- ArcFace R100 ONNX (my model): 0.6208
- ArcFace R100 MXNET: 0.7183

---

Both models on both approaches gave an accuracy of 0.5, but [this is somewhat expected](https://github.com/vitalwarley/research/issues/18#issuecomment-1146802531).

At first, I thought that the AUC gap was because of some reason that also caused a similar gap for LFW (as seen [here](https://github.com/vitalwarley/research/issues/18#issuecomment-1123782501)), but actually that was LFW accuracy -- with the onnx model, but my torch model gives the right accuracy. Instead, AUC on LFW was 0.99 for all approaches and models.

**Comment by vitalwarley** [comment]
Added Torch model to validation tests, but no luck getting right AUC yet.

## On Datamodule
![roc](https://user-images.githubusercontent.com/6365065/172073560-5996de4a-c31f-412c-86ff-e0ea21377470.png)
![CompareModelEmbeddings](https://user-images.githubusercontent.com/6365065/172073569-9f96395c-446a-47ee-8a8e-4cf4cc30d835.png)

Embeddings follow similar distribution, but that alone doesn't prove anything.

**Comment by vitalwarley** [comment]
I forgot to load the pre-trained weights in the torch model. :)

![roc](https://user-images.githubusercontent.com/6365065/172073906-b9f1c859-436a-45f3-a4c7-7f79f9ff785b.png)

**Comment by vitalwarley** [comment]
I retested on LFW and got the same results as [previously](https://github.com/vitalwarley/research/issues/18#issuecomment-1123782501). 👍🏼

**Comment by vitalwarley** [comment]
![CompareModelSimilarity](https://user-images.githubusercontent.com/6365065/172074424-1492dcda-4a0c-463e-baaf-779e9dc4430c.png)

**Comment by vitalwarley** [comment]
- [ ] Plot similarities by label for each model

**Comment by vitalwarley** [comment]
In val.py I use `PretrainModel`, which computes LFW accuracy based on pair distances instead of similarity

https://github.com/vitalwarley/research/blob/430039b3817893c5ea3c565b7c4c1873115236b4/model.py#L457-L485

Also, we used a 10-Fold validation scheme as did insightface implementation. I updated it to show the mean threshold for the given accuracy: 1.25. This makes sense if we look at the distance histograms for each pair label

![2022-06-11-111755_631x919_scrot](https://user-images.githubusercontent.com/6365065/173191649-abaa1391-1b05-420e-ad2b-3bad249ee34e.png)

Besides that, in val.py I use original images and its flipped version. How does that influence the results, if it really does?

Why computing AUC with similarities doesn't give bad results as well?

## Others findings

- CompareModelTorch embeddings range values were in (-5, 5) because #10. What does that implies in normalization? I don't think that is the problem because CompareModelMXNet has the same range.
- For cosine similiraity the vector magnitude doesn't matter.

**Comment by vitalwarley** [comment]
## LFW

Aparentemente o problema era na computação da acurácia.

![2022-06-11-213438_1945x1327_scrot](https://user-images.githubusercontent.com/6365065/173209378-e77655d4-a3f9-4f8e-b3ba-973820e57a3b.png)
![2022-06-11-213415_1319x1107_scrot](https://user-images.githubusercontent.com/6365065/173209379-34165118-d0d4-453a-924a-4f8ca6464227.png)

AUC continuou em .9985 e .9992 para Torch e MXNet, respectivamente. 

Como os resultados acima foram usando datamodule, então acredito que o processamento das imagens está correto.

**Comment by vitalwarley** [comment]
Nos gráficos acima eu usei `global_step=0` como eixo das distâncias/similaridades dos pares negativos.

**Comment by vitalwarley** [comment]
## FIW

### Distance/similarity histogram

That's odd. Maybe something wrong with `KinshipDataModule` and its datasets? Nonetheless, for mxnet we get AUC right, but not accuracy, as it should be .70 according with the paper.

![2022-06-11-215238_1434x1200_scrot](https://user-images.githubusercontent.com/6365065/173209712-c962b79a-1885-430f-b40e-17dbdb94cf8e.png)

## PR Curve

![2022-06-11-215158_2278x1314_scrot](https://user-images.githubusercontent.com/6365065/173209715-b86a4a7c-f9ab-4b28-8849-57e0e76d6147.png)

## Distance/similarity accuracy vs thresholds plot

![2022-06-11-215144_2230x1354_scrot](https://user-images.githubusercontent.com/6365065/173209718-5b462d36-d00a-4293-a2e1-da7c5a60f63f.png)

## AUC

Same as before:  0.6208 and 0.7183 for torch and mxnet, respectively.

![2022-06-11-215105_727x359_scrot](https://user-images.githubusercontent.com/6365065/173209720-e0cfab6f-7f76-466e-a1a8-16001d627768.png)

**Comment by vitalwarley** [comment]
I mistakenly swap ROC for PR curve. Soon I will add it again.

**Comment by vitalwarley** [comment]
After experimenting with FIW with and without datamodule, I suppose there is no difference in how images are processed. They give the same results and we can see it in the histograms below.

![2022-06-12-132237_1346x680_scrot](https://user-images.githubusercontent.com/6365065/173242975-ff9b29f0-36c8-4e7c-bc03-bcc6b2422efd.png)

The difference in the histograms may not be something that causes my torch model to not give 0.71 AUC, because we saw [before](https://github.com/vitalwarley/research/issues/18#issuecomment-1153032129) that histograms for LFW also have small differences between models.

What is it, then?

**Comment by vitalwarley** [comment]
I added tensorboard summaries in the original Shadrikov verification script to compare its similarities/distances with what I did

## Mine

![2022-06-12-163009_1589x662_scrot](https://user-images.githubusercontent.com/6365065/173250282-c884aeec-a24c-4ce4-b68d-0164043aee6f.png)

## Shadrikov

![2022-06-12-162935_1255x709_scrot](https://user-images.githubusercontent.com/6365065/173250299-6f9c9ea8-7c4b-477f-a0dd-ad03f9a25042.png)

---

From the plots above and [these](https://github.com/vitalwarley/research/issues/18#issuecomment-1153226395), I think we can remove the RAW scheme from the script. 

---

I also saved the embeddings from both models to try to understand them better, but I think that is inviable and impractical. I will try to create a script to inspect them in a better way.

**Comment by vitalwarley** [comment]
![2022-06-16-205110_1619x1308_scrot](https://user-images.githubusercontent.com/6365065/174196271-777125a4-259b-4be7-a124-6cb31ad67c50.png)

**Comment by vitalwarley** [comment]
I used the [arcface_torch](https://github.com/deepinsight/insightface/tree/master/recognition/arcface_torch) as a reference, but Shadrikov used mxnet directly. So I went to the [arcface_mxnet](https://github.com/deepinsight/insightface/tree/master/recognition/arcface_mxnet) implementation to see how things are really done.

## Image pre-processing

```python
        data = mx.sym.identity(data=data, name='id')
        data = data - 127.5
        data = data * 0.0078125
# https://github.com/deepinsight/insightface/blob/e78eee59caff3515a4db467976cf6293aba55035/recognition/arcface_mxnet/symbol/fresnet.py#L1057-L1060
```

This puts the image in approximately the (-1, 1) range. Is this equivalent to `(value - 0.5) / 0.5`, which is what I do in the transforms? I think so.

## FC Layer: definition

```python
        body = mx.sym.BatchNorm(data=body,
                                fix_gamma=False,
                                eps=2e-5,
                                momentum=bn_mom,
                                name='bn1')
        body = mx.symbol.Dropout(data=body, p=0.4)
        fc1 = mx.sym.FullyConnected(data=body,
                                    num_hidden=num_classes,
                                    name='pre_fc1')
        fc1 = mx.sym.BatchNorm(data=fc1,
                               fix_gamma=True,
                               eps=2e-5,
                               momentum=bn_mom,
                               name='fc1')
# https://github.com/deepinsight/insightface/blob/e78eee59caff3515a4db467976cf6293aba55035/recognition/arcface_mxnet/symbol/symbol_utils.py#L62-L76
```

Then we have this `fc1` layer above. One thing to note is the `dropout_rate=0.4` there. I didn't use dropout, but I don't think that's the reason for the gap in my results compared to Shadrikov's. The arcface_torch used p=0.

Configs:
- https://github.com/deepinsight/insightface/blob/e78eee59caff3515a4db467976cf6293aba55035/recognition/arcface_mxnet/sample_config.py#L181-L202
- https://github.com/deepinsight/insightface/blob/e78eee59caff3515a4db467976cf6293aba55035/recognition/arcface_mxnet/sample_config.py#L142-L147 

## FC Layer: normalization
```python
    elif config.loss_name == 'margin_softmax':
        _weight = mx.symbol.Variable("fc7_weight",
                                     shape=(config.num_classes,
                                            config.emb_size),
                                     lr_mult=config.fc7_lr_mult,
                                     wd_mult=config.fc7_wd_mult,
                                     init=mx.init.Normal(0.01))
        s = config.loss_s
        _weight = mx.symbol.L2Normalization(_weight, mode='instance')
        nembedding = mx.symbol.L2Normalization(
            embedding, mode='instance', name='fc1n') * s
        fc7 = mx.sym.FullyConnected(data=nembedding,
                                    weight=_weight,
                                    no_bias=True,
                                    num_hidden=config.num_classes,
                                    name='fc7')
        if config.loss_m1 != 1.0 or config.loss_m2 != 0.0 or config.loss_m3 != 0.0:
            if config.loss_m1 == 1.0 and config.loss_m2 == 0.0:
                s_m = s * config.loss_m3
                gt_one_hot = mx.sym.one_hot(gt_label,
                                            depth=config.num_classes,
                                            on_value=s_m,
                                            off_value=0.0)
                fc7 = fc7 - gt_one_hot
            else:
                zy = mx.sym.pick(fc7, gt_label, axis=1)
                cos_t = zy / s
                t = mx.sym.arccos(cos_t)
                if config.loss_m1 != 1.0:
                    t = t * config.loss_m1
                if config.loss_m2 > 0.0:
                    t = t + config.loss_m2
                body = mx.sym.cos(t)
                if config.loss_m3 > 0.0:
                    body = body - config.loss_m3
                new_zy = body * s
                diff = new_zy - zy
                diff = mx.sym.expand_dims(diff, 1)
                gt_one_hot = mx.sym.one_hot(gt_label,
                                            depth=config.num_classes,
                                            on_value=1.0,
                                            off_value=0.0)
                body = mx.sym.broadcast_mul(gt_one_hot, diff)
                fc7 = fc7 + body
# https://github.com/deepinsight/insightface/blob/e78eee59caff3515a4db467976cf6293aba55035/recognition/arcface_mxnet/train.py#L127-L170
```

The [parallel version](https://github.com/deepinsight/insightface/blob/e78eee59caff3515a4db467976cf6293aba55035/recognition/arcface_mxnet/train_parall.py#L143-L176) is equivalent.

All the above is done in the [ArcFace Loss](https://kevinmusgrave.github.io/pytorch-metric-learning/losses/#arcfaceloss) module. However, in mxnet implementation, the last layer in the inference graph is modified, while in my implementation, that doesn't happen.

https://github.com/vitalwarley/research/blob/4b4af298937ca94787ba7205795f72199baa20cd/model.py#L242-L264

This is equivalent to the arcface_torch implementation. They normalize weights and embeddings at inference.

```pythoon
      with torch.cuda.amp.autocast(self.fp16):
            norm_embeddings = normalize(embeddings)
            norm_weight_activated = normalize(self.weight_activated)
            logits = linear(norm_embeddings, norm_weight_activated)
        if self.fp16:
            logits = logits.float()
        logits = logits.clamp(-1, 1)


        logits = self.margin_softmax(logits, labels)
        loss = self.dist_cross_entropy(logits, labels)
        return loss
```
## Conclusions

- I think the embeddings I return from my model are proper because it seems that both arcface_mxnet and arcface_torch normalize the embeddings and weights at training, which is what I do with ArcFace Loss.
  - So [these](https://github.com/vitalwarley/research/issues/18#issuecomment-1158288045) histograms prove that... I think.

## Next steps

- Implement the other schemes that Shadrikov used: classification, classification + normalization.
- Train model with arcface_torch and compare it with Shadrikov results and mine.

**Comment by vitalwarley** [comment]
> Train model with arcface_torch and compare it with Shadrikov results and mine.

No need for that. I tried to use the `arcface_torch` pre-trained model again. It seems that this model is bigger, because I couldn't use bs=128 like my model, but instead only bs=32.

![2022-07-04-084354_1435x320_scrot](https://user-images.githubusercontent.com/6365065/177148001-4e55dce4-789a-44d0-92bb-efb4d8ec160d.png)

- fiw/1655425397 is before (16/06). fiw/1656933605 is today. See that the overlapping area is less than before.
- Darker blue represents the positive samples.

Also, compare with mxnet

![2022-07-04-084639_669x325_scrot](https://user-images.githubusercontent.com/6365065/177148586-7dee7e94-d85d-4338-92e7-7f42f09f72d8.png)

Basically the same distribution.

AUC was 0.7198, which is very good because previously it was 0.6208 with my model. The accuracy went from 0.59 to 0.66 -- same value as mxnet now. We can conclude two things: 1) something is wrong with my model and/or its training; 2) Shadrikov's accuracy seems to be really 0.66, not 0.70, as in the paper. 

> That's odd. Maybe something wrong with KinshipDataModule and its datasets? Nonetheless, for mxnet we get AUC right, but not accuracy, as it should be .70 according with the paper.
> https://github.com/vitalwarley/research/issues/18#issuecomment-1153034666

This is my fault, though. That accuracy seems to be the weighted average from the accuracies for each relation type. I will confirm that with #19.

**Comment by vitalwarley** [comment]
How different is the inference with insightface vs my pretrained model?

https://github.com/vitalwarley/research/blob/b6097da27bf63989ee28bf0faead85d50ba54701/model.py#L113-L134

and

https://github.com/vitalwarley/research/blob/b6097da27bf63989ee28bf0faead85d50ba54701/model.py#L245-L255

![2022-07-04-090222_1469x575_scrot](https://user-images.githubusercontent.com/6365065/177150803-67711960-419f-4432-bdc6-0e8dc5f1af52.png)

It seems to me that my model embeddings distribution (pink) is less dispersed compared to insightface.

**Comment by vitalwarley** [comment]
I evaluate with the checkpoint from the 49th epoch. Before, it was from the 34th epoch. AUC increased by 0.0167 points and ACC by 0.01. 

I used a batch size of 2048 (1024 * 2 because of accumulated grad batches) for two GPUs, therefore the total batch size was 4096. Insightface training was with a bs=128 for 8 GPUs (total bs = 1024). I did it for 50 epochs, they did it for 20 epochs.

**Comment by vitalwarley** [comment]
After #15 and #13, I improved my results. AUC on FIW went to 0.6717 and accuracy went to 0.63. They are less than Shadrikov's model: 0.7183 and 0.67, respectively.

**Comment by vitalwarley** [comment]
I think two things remain to be tested, but not now because I will focus on [these issues](https://github.com/vitalwarley/research/issues/18#issuecomment-1177986568).

1. Use insightface's resnet101 instead of timm's.
2. Use PartialFC (#7).

I couldn't find any other explanation. My pieces of evidence are: 1) insightface's PyTorch backbone gives the [same results](https://github.com/vitalwarley/research/issues/18#issuecomment-1173724030) as the MXNet; 2) #13 -- learning rate curve and loss curve are similar to insightface's.
